### ISSUE TITLE
Question about evaluation batch_size and positive_u

### ISSUE THREAD

**Issue by v1xerunt** [RUCAIBox/RecBole#1843]
**Describe the bug**
Hi, I'm working on a custom dataset which has 4255 users and 24525 items and the dataset is split into train:val:test = 8:1:1. I'm using a benchmark setting where all users in the test set are new users. However, I'm confused about the inconsistency issue with the test dataloader. The batch_size in the yaml file is set to 2048. 

**Q1.** However, the `test_data._batch_size` gives me 24525, which is the item number of the dataset and `test_data.step` gives me 1. I understand the batch size is recalculated, however, the `len(test_data)` gives me 425, which is the user number in the test set. I assume `len(test_data)` should equal to `test_data.step`?

**Q2.** I took a deeper look into the batch data of the test set. I found all `positive_u` in each iteration is a zero vector while `positive_i` is correct (please see my output below). I wonder if this is expected or if there is something wrong with the code or settings.

Thank you very much.

**To Reproduce**
1. YAML file
```
data_path: '.\atomic\'
checkpoint_dir: '.\models\'
train_neg_sample_args:
  uniform: 1
benchmark_filename: ['train', 'val', 'test']

additional_feat_suffix: [userembd, itemembd]
load_col:
    inter: [user_id, item_id]
    # inter/user/item/...: As usual
    userembd: [uid, user_emb]
    itemembd: [iid, item_emb]
alias_of_user_id: [uid]
alias_of_item_id: [iid]
preload_weight:
    uid: user_emb
    iid: item_emb

embedding_size: 64
n_layers: 2
reg_weight: 0.00001

epochs: 300
train_batch_size: 2048
learning_rate: 0.01
eval_batch_size: 2048
topk: 10
```

2. Code
```
from logging import getLogger
from recbole.utils import init_logger, init_seed
from recbole.trainer import Trainer
from custom.light_gcn import LightGCN
from recbole.config import Config
from recbole.data import create_dataset, data_preparation
import torch

if __name__ == '__main__':

    config = Config(model=LightGCN, dataset='trial_zero', config_file_list=["D:\OneDrive - University of Edinburgh\paper\[01 PRE] Trial Recomm\\atomic\\lightgcn_custom.yaml"])
    init_seed(config['seed'], config['reproducibility'])

    # logger initialization
    init_logger(config)
    logger = getLogger()

    logger.info(config)

    # dataset filtering
    dataset = create_dataset(config)
    logger.info(dataset)
    
    
    # dataset splitting
    train_data, valid_data, test_data = data_preparation(config, dataset)

    print('len(test_data): ', len(test_data))
    print('config batch_size: ', config['train_batch_size'])
    print('test _batch_size: ', test_data._batch_size)
    print('test step: ', test_data.step)
    
    for i, (_, _, pos_u, pos_i) in enumerate(test_data):
        print('Pos u @ iter %d: ' % i, pos_u)
        print('Pos i @ iter %d: ' % i, pos_i)
        if i == 5:
            break
```

3. Output
```
General Hyper Parameters:
gpu_id = 0
use_gpu = True
seed = 2020
state = INFO
reproducibility = True
data_path = .\trial_zero
checkpoint_dir = .\
show_progress = True
save_dataset = False
dataset_save_path = None
save_dataloaders = False
dataloaders_save_path = None
log_wandb = False

Training Hyper Parameters:
epochs = 300
train_batch_size = 2048
learner = adam
learning_rate = 0.01
train_neg_sample_args = {'uniform': 1, 'distribution': 'uniform', 'sample_num': 1, 'alpha': 1.0, 'dynamic': False, 'candidate_num': 0}
eval_step = 1
stopping_step = 10
clip_grad_norm = None
weight_decay = 0.0
loss_decimal_place = 4

Evaluation Hyper Parameters:
eval_args = {'split': {'RS': [0.8, 0.1, 0.1]}, 'group_by': 'user', 'order': 'RO', 'mode': 'full'}
repeatable = False
metrics = ['Hit']
topk = [10]
valid_metric = MRR@10
valid_metric_bigger = True
eval_batch_size = 2048
metric_decimal_place = 4

Dataset Hyper Parameters:
field_separator = 	
seq_separator =  
USER_ID_FIELD = user_id
ITEM_ID_FIELD = item_id
RATING_FIELD = rating
TIME_FIELD = timestamp
seq_len = None
LABEL_FIELD = label
threshold = None
NEG_PREFIX = neg_
load_col = {'inter': ['user_id', 'item_id'], 'userembd': ['uid', 'user_emb'], 'itemembd': ['iid', 'item_emb']}
unload_col = None
unused_col = None
additional_feat_suffix = ['userembd', 'itemembd']
rm_dup_inter = None
val_interval = None
filter_inter_by_user_or_item = True
user_inter_num_interval = [0,inf)
item_inter_num_interval = [0,inf)
alias_of_user_id = ['uid']
alias_of_item_id = ['iid']
alias_of_entity_id = None
alias_of_relation_id = None
preload_weight = {'uid': 'user_emb', 'iid': 'item_emb'}
normalize_field = None
normalize_all = None
ITEM_LIST_LENGTH_FIELD = item_length
LIST_SUFFIX = _list
MAX_ITEM_LIST_LENGTH = 50
POSITION_FIELD = position_id
HEAD_ENTITY_ID_FIELD = head_id
TAIL_ENTITY_ID_FIELD = tail_id
RELATION_ID_FIELD = relation_id
ENTITY_ID_FIELD = entity_id
benchmark_filename = ['train', 'val', 'test']

Other Hyper Parameters: 
worker = 0
wandb_project = recbole
shuffle = True
require_pow = False
enable_amp = False
enable_scaler = False
transform = None
embedding_size = 64
n_layers = 2
reg_weight = 1e-05
numerical_features = []
discretization = None
kg_reverse_r = False
entity_kg_num_interval = [0,inf)
relation_kg_num_interval = [0,inf)
MODEL_TYPE = ModelType.GENERAL
MODEL_INPUT_TYPE = InputType.PAIRWISE
eval_type = EvaluatorType.RANKING
single_spec = True
local_rank = 0
device = cuda
eval_neg_sample_args = {'distribution': 'uniform', 'sample_num': 'none'}


24 Aug 05:48    INFO  trial_zero
The number of users: 4255
Average actions of users: 15.768941176470587
The number of items: 24525
Average actions of items: 2.7327515902789106
The number of inters: 67018
The sparsity of the dataset: 99.93577813952764%
Remain Fields: ['user_id', 'item_id', 'uid', 'user_emb', 'iid', 'item_emb']
24 Aug 05:48    INFO  [Training]: train_batch_size = [2048] train_neg_sample_args: [{'uniform': 1, 'distribution': 'uniform', 'sample_num': 1, 'alpha': 1.0, 'dynamic': False, 'candidate_num': 0}]
len(test_data):  425
config batch_size:  2048
24 Aug 05:48    INFO  [Evaluation]: eval_batch_size = [2048] eval_args: [{'split': {'RS': [0.8, 0.1, 0.1]}, 'group_by': 'user', 'order': 'RO', 'mode': 'full'}]
test _batch_size:  24525
test step:  1
Pos u @ iter 0:  tensor([0, 0, 0])
Pos i @ iter 0:  tensor([ 2128,  6042, 21341])
Pos u @ iter 1:  tensor([0, 0])
Pos i @ iter 1:  tensor([16532, 15767])
Pos u @ iter 2:  tensor([0, 0, 0, 0])
Pos i @ iter 2:  tensor([1660, 8428, 9052, 2349])
Pos u @ iter 3:  tensor([0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0])
Pos i @ iter 3:  tensor([  833, 12548,  6887,  6925,  5967, 23824,  5969,  9010,  5972, 13146,
        12541])
Pos u @ iter 4:  tensor([0, 0, 0, 0, 0])
Pos i @ iter 4:  tensor([19521,  5187,  9226, 20241, 23825])
Pos u @ iter 5:  tensor([0])
Pos i @ iter 5:  tensor([3440])
```

**Comment by Sherry-XLL** [comment]
Hello @v1xerunt, thanks for your attention to RecBole!

For your first question, it is related to the evaluation mode you have set. Under the setting of full evaluation (`'mode': 'full'`), what the recommender system needs to do is a recall task. It means that for each user, we need to select the _k_ items that the user is most interested in from all candidate items, that is, the top-_k_ recommendation evaluation. Since we need to calculate the predicted scores of each user and all candidate items, the `batch_size` will be adapted to the total number of items, so that the calculations of the same user are within one batch. After the batch adaptation, the size of the batch is the number of users in the test set.

For your second question, this is not a bug in our code, the specific implementation is related to the underlying code logic. When calculating predictions, we will calculate the scores for each user and all items in each batch as a matrix named `scores_tensor`, with a shape of `batch_size * item_num`. `positive_u` and `positive_i` represent the horizontal and vertical coordinates of positive items in `scores_tensor`, respectively. Due to your settings, there is only one user in each batch, the size of `scores_tensor` is `1 * item_num`. Therefore, all abscissa values are `0`, making it convenient for us to directly index the scoring matrix.

Thanks again for your feedback, and feel free to contact us if you have further questions.

**Comment by v1xerunt** [comment]
Thank you very much for your clarifications. It makes sense to me now.
